packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp

```diff
@@ -0,0 +1,178 @@
+    defaultName: boolean;
+  }
+
+  model JsonClientNameAndEncodedNameModel {
+    @doc("Pass in true")
+    @encodedName("application/json", "wireName")
+    defaultName: boolean;
+  }
+
+  model ClientClientNameAndEncodedNameModel {
+    @doc("Pass in true")
+    @clientName("clientName")
+    defaultName: boolean;
+  }
+
+  model JsonAndClientClientNameAndEncodedNameModel {
+    @doc("Pass in true")
+    @clientName("clientName")
+    @encodedName("application/json", "wireName")
+    defaultName: boolean;
+  }
+
+  @scenario
+  @scenarioDoc("""
+  Testing that we can project the JSON name on the wire from defaultName -> wireName.
+  Your generated SDK should generate JsonClientNameAndEncodedNameModel with one property `defaultName` with wire name `wireName`.
+
+  Expected request body:
+  ```json
+  {"wireName": true}
+  ```
+  """)
+  @route("/json")
+  @post
+  op json(...JsonClientNameAndEncodedNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the client name in our generated SDKs.
+Your generated SDK should generate ClientClientNameAndEncodedNameModel with one property `clientName` with wire name `defaultName`.
+
+Expected request body:
+```json
+{"defaultName": true}
+```
+""")
+  @route("/client")
+  @post
+  op client(...ClientClientNameAndEncodedNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the language specific name in our generated SDKs.
+Your generated SDK should generate ClientClientNameAndEncodedNameModel with one property with your language specific property name and wire name `defaultName`.
+
+Expected request body:
+```json
+{"defaultName": true}
+```
+""")
+  @route("/language")
+  @post
+  op language(...LanguageClientNameAndEncodedNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the client name and the wire name.
+Your generated SDK should generate JsonAndClientClientNameAndEncodedNameModel with one property with client name `clientName` and wire name `wireName`.
+
+Expected request body:
+```json
+{"wireName": true}
+```
+""")
+  @route("/json-and-client")
+  @post
+  op jsonAndClient(...JsonAndClientClientNameAndEncodedNameModel): NoContentResponse;
+}
+
+// May add query/header project after https://github.com/microsoft/cadl/issues/1487
+
+@scenario
+@scenarioDoc("""
+Testing that we can project the operation name.
+Your generated SDK should generate an operation called `clientName`.
+
+Expected status code: 204
+""")
+@route("/operation")
+@clientName("clientName")
+@post
+op operation(): NoContentResponse;
+
+@scenario
+@scenarioDoc("""
+Testing that we can project a parameter name.
+Your generated SDK should generate an operation `parameter` with a single parameter called `clientName`.
+
+Expected query parameter: `default-name="true"`
+
+""")
+@route("/parameter")
+@post
+op parameter(
+  @clientName("clientName")
+  @query
+  `default-name`: string,
```

Comment:
Can we add test for query/header wire name renaming?

Reply:
added here https://github.com/Azure/cadl-ranch/pull/511/files#diff-64d6117e7c3da950e6cf3f1b07c36975dc89959e0848cb1da15af7a2e8993b20R116-R149

Reply:
It seems that `@encodedName` can not be apply to query or header parameter. See [typespec playground link](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwOwpAc2VydmljZSh7CiAgdGl0bGU6ICJXaWRnZXQgU8YbIiwKICB2ZXJzaW9uOiAiMS4wLjAiLAp9KQpuYW1lc3BhY2UgRGVtb8cvOwoKQHJvdXRlKCIvcXVlcnkiKQpAcG9zdApvcCDFESgKICBAZW5jb2RlZE5hbWUoImFwcGxpY2F0aW9uL2pzb24iLCAixSstxGMiKcQxxRAKICBkZWZhdWx0xDk6IHN0cmluZywKKTogTm9Db250ZW50UmVzcG9uc%2B0AiWhlYWRlcuwAisYS%2FwCL5wCLxizrAIzGEf8Aje8AjQ%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

Reply:
Yeah it doesn’t apply to query params as those are not json things. Not sure if we wanted to support that what is the right media type for it if any. To change the name of header, query you can use the param of the corresponding decorators

Reply:
I have removed the encodedName for query and header parameter and header in response.